internal/server/v2/observation/direct.go

```diff
@@ -48,70 +49,123 @@ func FetchDirect(
+				series := btData[entity][variable].SourceSeries
+				if len(series) > 0 {
+					// Sort series by rank
+					sort.Sort(ranking.SeriesByRank(series))
+					for _, series := range series {
+						facet := util.GetFacet(series)
+						if filter != nil && filter.Domain != "" {
+							url, err := url.Parse(facet.ProvenanceUrl)
+							if err != nil {
+								return nil, err
+							}
+							// To match domain or subdomain. For example, a provenance url of
+							// abc.xyz.com can match filter "xyz.com" and "abc.xyz.com".
+							if !strings.HasSuffix(url.Hostname(), filter.Domain) {
+								continue
+							}
 						}
-					}
-					facetID := util.GetFacetID(facet)
-					obsList := []*pb.PointStat{}
-					for date, value := range series.Val {
-						ps := &pb.PointStat{
-							Date:  date,
-							Value: proto.Float64(value),
+						facetID := util.GetFacetID(facet)
+						obsList := []*pb.PointStat{}
+						for date, value := range series.Val {
+							ps := &pb.PointStat{
+								Date:  date,
+								Value: proto.Float64(value),
+							}
+							if queryDate != "" && queryDate != LATEST && queryDate != date {
+								continue
+							}
+							obsList = append(obsList, ps)
 						}
-						if queryDate != "" && queryDate != LATEST && queryDate != date {
+						if len(obsList) == 0 {
 							continue
 						}
-						obsList = append(obsList, ps)
-					}
-					if len(obsList) == 0 {
-						continue
-					}
-					sort.SliceStable(obsList, func(i, j int) bool {
-						return obsList[i].Date < obsList[j].Date
-					})
-					if queryDate == LATEST {
-						obsList = obsList[len(obsList)-1:]
-						// If there is higher quality series, then do not pick from the inferior
-						// facet even it could have more recent data.
-						if len(entityObservation.OrderedFacets) > 0 && stat.IsInferiorFacetPb(series) {
-							break
+						sort.SliceStable(obsList, func(i, j int) bool {
+							return obsList[i].Date < obsList[j].Date
+						})
+						if queryDate == LATEST {
+							obsList = obsList[len(obsList)-1:]
+							// If there is higher quality series, then do not pick from the inferior
+							// facet even it could have more recent data.
+							if len(entityObservation.OrderedFacets) > 0 && stat.IsInferiorFacetPb(series) {
+								break
+							}
 						}
+						result.Facets[facetID] = facet
+						entityObservation.OrderedFacets = append(
+							entityObservation.OrderedFacets,
+							&pbv2.FacetObservation{
+								FacetId:      facetID,
+								Observations: obsList,
+							},
+						)
 					}
-					result.Facets[facetID] = facet
-					entityObservation.OrderedFacets = append(
-						entityObservation.OrderedFacets,
-						&pbv2.FacetObservation{
-							FacetId:      facetID,
-							Observations: obsList,
-						},
-					)
 				}
+				result.ByVariable[variable].ByEntity[entity] = entityObservation
 			}
-			result.ByVariable[variable].ByEntity[entity] = entityObservation
+		}
+	} else if store.SQLiteClient != nil {
```

Comment:
nice.  Maybe could have a TODO to modularize the different backends a bit?

Reply:
Yea, that sounds good. I will do that after adding more APIs / need a new connector for Cloud SQL